### PR TITLE
fix: clear event binding function references

### PR DESCRIPTION
**Pull request by odedhutzler**
### Description of the Changes

Created an object to hold all the event handlers, so we can unbind it properly. 
Today we don't really unbind the handler because it's a new handler and not a reference to the previous one.

### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [x] test are passing in local environment 
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [x] Docs have been updated

### CODE REVIEW

File: src/dash-adapter.js [hunk 1]
```diff
@@ -53,6 +53,13 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
    * @private
    */
   _shaka: any;
+  /**
+   * an object containing all the events we bind and unbind to.
+   * @member {Object} - _adapterEventsBindings
+   * @type {Object}
+   * @private
+   */
+  _adapterEventsBindings: Object = {};
```
Comment (OrenMe) [inline src/dash-adapter.js]: flow type `{[name: string]: function}`

File: src/dash-adapter.js [hunk 2]
```diff
@@ -202,19 +210,29 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
     }
   }
 
+
+  _createAdapterFunctionBindings(): void{
+
```
Comment (OrenMe) [inline src/dash-adapter.js]: why empty line?

File: src/dash-adapter.js [hunk 3]
```diff
@@ -202,19 +210,29 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
...
+      this._adapterEventsBindings['error'] =  this._onError.bind(this);
```
Comment (OrenMe) [inline src/dash-adapter.js]: can we use arrow function?

Comment (OrenMe) [reply src/dash-adapter.js]: also, need to clear these bounded methods on destroy or we will have memory leaks possibly

File: src/dash-adapter.js (hunk 2 above)
Comment (odedhutzler) [reply src/dash-adapter.js]: there is no spoon (empty line).

File: src/dash-adapter.js [hunk 4]
```diff
@@ -202,19 +210,28 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
...
-    this._shaka.addEventListener('adaptation', this._onAdaptation.bind(this));
-    this._shaka.addEventListener('error', this._onError.bind(this));
-    this._shaka.addEventListener('buffering', this._onBuffering.bind(this));
+    this._shaka.addEventListener('adaptation', this._adapterEventsBindings.adaption);
+    this._shaka.addEventListener('error', this._adapterEventsBindings.error);
+    this._shaka.addEventListener('buffering', this._adapterEventsBindings.buffering);
     //TODO use events enum when available
-    this._videoElement.addEventListener('waiting', this._onWaiting.bind(this));
-    this._videoElement.addEventListener('playing', this._onPlaying.bind(this));
+    this._videoElement.addEventListener('waiting', this._adapterEventsBindings.waiting);
```
Comment (OrenMe) [inline src/dash-adapter.js]: Use ENUM for HTML5 events

File: src/dash-adapter.js [hunk 5]
```diff
@@ -202,19 +210,28 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
...
    * @function _addBindings
    * @private
    * @returns {void}
    */
   _addBindings(): void {
-    this._shaka.addEventListener('adaptation', this._onAdaptation.bind(this));
-    this._shaka.addEventListener('error', this._onError.bind(this));
-    this._shaka.addEventListener('buffering', this._onBuffering.bind(this));
+    this._shaka.addEventListener('adaptation', this._adapterEventsBindings.adaption);
+    this._shaka.addEventListener('error', this._adapterEventsBindings.error);
```
Comment (OrenMe) [inline src/dash-adapter.js]: Create local ENUM for Shaka event names

File: src/dash-adapter.js [hunk 6]
```diff
@@ -7,6 +7,22 @@ import {Error} from 'playkit-js'
...
+/**
+ * Shaka events enum
+ * @type {Object}
+ * @const
+ */
+const SHAKA_EVENTS: ShakaEventsType = {
+  ERROR: 'error' ,
+  ADAPTION: 'adaption',
+  BUFFERING: 'buffering',
+  WAITING: 'waiting',
```
Comment (OrenMe) [inline src/dash-adapter.js]: WAITING and PLAYING are HTML5 events

File: src/dash-adapter.js [hunk 7]
```diff
@@ -7,6 +7,20 @@ import {Error} from 'playkit-js'
 import Widevine from './drm/widevine'
 import PlayReady from './drm/playready'
 
+
+type ShakaEventsType = { [event: string]: string };
```
Comment (dan-ziv) [inline src/dash-adapter.js]: change to `ShakaEventType`

File: src/dash-adapter.js [hunk 8]
```diff
@@ -7,6 +7,20 @@ import {Error} from 'playkit-js'
...
+
+/**
+ * Shaka events enum
+ * @type {Object}
+ * @const
+ */
+const SHAKA_EVENTS: ShakaEventsType = {
```
Comment (dan-ziv) [inline src/dash-adapter.js]: Enums convention are to lead with capitals and refer as singular (i.e. `ShakaEvent`)

File: src/dash-adapter.js [hunk 9]
```diff
@@ -202,19 +224,27 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
     }
   }
 
+
+  _createAdapterFunctionBindings(): void {
```
Comment (dan-ziv) [inline src/dash-adapter.js]: We can do this logic statically when we declare the `_adapterEventsBindings ` variable, no need to create method for it

File: src/dash-adapter.js [hunk 10]
```diff
@@ -202,19 +224,27 @@ export default class DashAdapter extends BaseMediaSourceAdapter {
     }
   }
 
+
```
Comment (dan-ziv) [inline src/dash-adapter.js]: remove redundant lines (!!!)

File: src/dash-adapter.js (hunk 9 above)
Comment (OrenMe) [reply src/dash-adapter.js]: @dan-ziv what do you mean statically? we want to have the this bindings

Comment (dan-ziv) [reply src/dash-adapter.js]: _adapterEventsBindings: { [name: string]: Function } = {
    [ShakaEvent.ERROR] : (event) => this._onError(event),
    ...
  };

**Pull request merged**